pyro/infer/abstract_infer.py

```diff
@@ -2,133 +2,63 @@
-            else:
-                # Value has already been seen.
-                logits[ix] = log_sum_exp(torch.stack([logits[ix], logit]))
-
-        logits = torch.stack(logits).reshape(-1)
-        logits -= log_sum_exp(logits)
-        logits = logits - log_sum_exp(logits)
-        d = dist.Categorical(logits=logits)
-        return d, values
-
-    def _gen_weighted_samples(self, *args, **kwargs):
-        raise NotImplementedError("_gen_weighted_samples is abstract method")
-
-    def sample(self, *args, **kwargs):
-        sample_shape = kwargs.pop("sample_shape", None)
-        if sample_shape:
-            raise ValueError("Arbitrary `sample_shape` not supported by Histogram class.")
-        d, values = self._dist_and_values(*args, **kwargs)
-        ix = d.sample()
-        return values[ix]
-
-    __call__ = sample
-
-    def log_prob(self, val, *args, **kwargs):
-        d, values = self._dist_and_values(*args, **kwargs)
-        ix = _index(values, val)
-        return d.log_prob(torch.tensor([ix]))
-
-    def enumerate_support(self, *args, **kwargs):
-        d, values = self._dist_and_values(*args, **kwargs)
-        return values[:]
-
-
-class Marginal(Histogram):
-    """
-    :param trace_dist: a TracePosterior instance representing a Monte Carlo posterior
-
-    Marginal histogram distribution.
-    Turns a TracePosterior object into a Distribution
-    over the return values of the TracePosterior's model.
-    """
     def __init__(self, trace_dist, sites=None):
         assert isinstance(trace_dist, TracePosterior), \
             "trace_dist must be trace posterior distribution object"
 
         if sites is None:
-            sites = "_RETURN"
-
-        assert isinstance(sites, (str, list)), \
-            "sites must be either '_RETURN' or list"
-
-        if isinstance(sites, str):
-            assert sites in ("_RETURN",), \
-                "sites string must be '_RETURN'"
+            sites = ["_RETURN"]
 
         self.sites = sites
-        super(Marginal, self).__init__()
         self.trace_dist = trace_dist
+        self.empirical_marginal = None
+        super(Distribution, self).__init__()
+
+    def sample(self, *args, **kwargs):
+        sample_shape = kwargs.pop("sample_shape", torch.Size())
+        return self._get_marginal(*args, **kwargs).sample(sample_shape)
+
+    def log_prob(self, value, *args, **kwargs):
+        return self._get_marginal(*args, **kwargs).log_prob(value)
+
+    def enumerate_support(self, *args, **kwargs):
+        return self._get_marginal(*args, **kwargs).enumerate_support()
 
-    def _gen_weighted_samples(self, *args, **kwargs):
-        for tr, log_w in poutine.block(self.trace_dist._traces)(*args, **kwargs):
-            if self.sites == "_RETURN":
-                val = tr.nodes["_RETURN"]["value"]
-            else:
-                val = {name: tr.nodes[name]["value"]
-                       for name in self.sites}
-            yield (val, log_w)
+    @util.memoize
```

Comment:
We should probably remove this for non-`Search` marginals

Reply:
Currently, this code path is only called by `Search` in existing tests/examples. Non Search code can call it too, but they should ideally just call `TracePosterior.marginal()` directly, which will store the results internally. In any case, disregarding any memory issues related to memoizing, I think the logic should be fine but will work in a restricted way (i.e. only when args, kwargs are the same instances).